ocaml/xapi/xapi_clustering.ml

```diff
@@ -551,12 +553,51 @@ module Watcher = struct
     done ;
     Mutex.execute mu (fun () -> cluster_change_watcher := None)
 
+  let watch_cluster_stack_version ~__context ~host =
+    while !Daemon.enabled do
+      ( match find_cluster_host ~__context ~host with
+      | Some ch ->
+          let cluster_ref = Db.Cluster_host.get_cluster ~__context ~self:ch in
+          let cluster_rec =
+            Db.Cluster.get_record ~__context ~self:cluster_ref
+          in
+          if
+            Cluster_stack.of_version
+              ( cluster_rec.API.cluster_cluster_stack
+              , cluster_rec.API.cluster_cluster_stack_version
+              )
+            = Cluster_stack.Corosync2
+          then
+            debug "%s: Detected corosync 2 running as cluster stack"
+              __FUNCTION__ ;
+          let body =
+            "The current cluster stack version of Corosync 2 is out of date, \
+             consider updating to Corosync 3"
+          in
+          let name, priority = Api_messages.cluster_stack_out_of_date in
+          let host_uuid = Db.Host.get_uuid ~__context ~self:host in
+
+          Helpers.call_api_functions ~__context (fun rpc session_id ->
+              ignore
+              @@ Client.Client.Message.create ~rpc ~session_id ~name ~priority
+                   ~cls:`Host ~obj_uuid:host_uuid ~body
+          )
```

Comment:
Looks like we are missing `()` to include this in the `then` block.

Also, if we have detected Corosync 3, then we can just exit the thread?

Reply:
Yes, we can do that as well. I was thinking about sending periodic notifications to the user urging them to do the upgrade, which has a 7-day period

Reply:
With this code, a `cluster_stack_out_of_date` message is sent even if the cluster stack is _not_ out of date.

Reply:
Yes, you are absolutely right, I am missing a bracket around the if statement